geonode/geoserver/geofence.py

```diff
@@ -310,73 +340,127 @@ def insert_rule(self, rule: Rule):
+                )
                 raise GeofenceException(f"Error while running batch {batch.log_name}: [{r.status_code}]")
 
-            return
+            return True
 
         except Exception as e:
-            logger.debug(f"Error while requesting batch execution {batch.log_name}", exc_info=e)
+            logger.info(f"Error while requesting batch exec {batch.log_name}")
+            logger.debug(f"Error while requesting batch exec {batch.log_name} --> {batch.get_object()}", exc_info=e)
             raise GeofenceException(f"Error while requesting batch execution {batch.log_name}: {e}")
 
-    def purge_all_rules(self):
+
+class GeoFenceUtils:
+    def __init__(self, client: GeoFenceClient):
+        self.geofence = client
+
+    def delete_all_rules(self):
         """purge all existing GeoFence Cache Rules"""
-        rules_objs = self.get_rules()
+        rules_objs = self.geofence.get_rules()
         rules = rules_objs["rules"]
 
         batch = Batch("Purge All")
         for rule in rules:
             batch.add_delete_rule(rule["id"])
 
         logger.debug(f"Going to remove all {len(rules)} rules in geofence")
-        self.run_batch(batch)
-
-    def purge_layer_rules(self, layer_name: str, workspace: str = None):
-        """purge existing GeoFence Cache Rules related to a specific Layer"""
-        gs_rules = self.get_rules(workspace=workspace, workspace_any=False, layer=layer_name, layer_any=False)
-
-        batch = Batch(f"Purge {workspace}:{layer_name}")
-
-        if gs_rules and gs_rules["rules"]:
-            logger.debug(f"Going to remove {len(gs_rules['rules'])} rules for layer '{layer_name}'")
-            for r in gs_rules["rules"]:
-                if r["layer"] and r["layer"] == layer_name:
-                    batch.add_delete_rule(r["id"])
-                else:
-                    logger.debug(f"Bad rule retrieved for dataset '{layer_name}': {r}")
-        self.run_batch(batch)
+        self.geofence.run_batch(batch)
+
+    def collect_delete_layer_rules(self, workspace_name: str, layer_name: str, batch: Batch = None) -> Batch:
+        """Collect delete operations in a Batch for all rules related to a layer"""
+
+        try:
+            # Scan GeoFence Rules associated to the Dataset
+            gs_rules = self.geofence.get_rules(
+                workspace=workspace_name, workspace_any=False, layer=layer_name, layer_any=False
+            )
+
+            if not batch:
+                batch = Batch(f"Delete {workspace_name}:{layer_name}")
+
+            cnt = 0
+            if gs_rules and gs_rules["rules"]:
+                logger.debug(
+                    f"Going to collect {len(gs_rules['rules'])} rules for layer '{workspace_name}:{layer_name}'"
+                )
+                for r in gs_rules["rules"]:
+                    if r["layer"] and r["layer"] == layer_name:
+                        batch.add_delete_rule(r["id"])
+                        cnt += 1
```

Comment:
Isn't the count provided by the `batch.length()` as soon as all the rules are added?

Reply:
No: if the delete operations are collected in an existing batch, lenght would be > than the delete operations inserted